### PR TITLE
expose information about oversized images

### PR DESCRIPTION
**Pull request by devoncarew**
expose information about oversized images:
- add a toggle for the `ext.flutter. invertOversizedImages` service extension; this is on the Inspector page right now - it seems more immediately something people would use while looking at their app that a memory related toggle. Happy to move to to wherever is most useful however
- process the new `Flutter.ImageSizesForFrame` events; these include information about how much more memory an image is using than it needs
- handle displaying Flutter.Error event information on the logging page. Previously we had a todo here to render using the Inspector mechanisms. Since this was not implemented however, the user just saw a blank details area. This instead renders as JSON; we can improve this going forward.

@terrylucas and @jacob314 for review and comments

@dnfield - the default value for `debugImageOverheadAllowance` seems super twitchy. I think we should likely dial it back so that the flutter image errors trigger less often. I _believe_ it's currently on overage of 1024 bytes. Perhaps some value like 2x memory usage?

Screen Shot 2020-11-17 at 1 41 45 PM

Screen Shot 2020-11-17 at 1 42 17 PM

Screen Shot 2020-11-17 at 1 43 24 PM

Screen Shot 2020-11-17 at 1 44 01 PM

### CODE REVIEW

**Comment by dnfield** [comment]
I'd be fine with changing the default, but I think using a multiplier is going to be misleading.

For example, if you are displaying a 48x48 icon with an image that's 100x100, you may not care, but if you're displaying a full screen image that's taking 1.9x more memory you should probably care.

**Comment by devoncarew** [comment]
> I'd be fine with changing the default, but I think using a multiplier is going to be misleading.
> 
> For example, if you are displaying a 48x48 icon with an image that's 100x100, you may not care, but if you're displaying a full screen image that's taking 1.9x more memory you should probably care.

Would a fixed amount of inefficient memory usage work? 128kb / image?

**Comment by devoncarew** [comment]
> DBC based on screenshot - the icon for the service extension toggle button should be changed from the debug banner icon

I did take a quick look through our existing icons in `icons/` but nothing jumped out at me; I'll take a look again.

File: packages/devtools_app/lib/src/console.dart [hunk 1]
```diff
@@ -53,6 +53,45 @@ class Console extends StatelessWidget {
...
+  Widget build(BuildContext context) {
+    return Column(
+      children: [
+        if (title != null) title,
+        Expanded(
+          child: Material(
+            child: Stack(
+              fit: StackFit.expand,
+              children: [
+                _ConsoleText(text: textContent),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/console.dart:81]: why isn't this a list view where each element is a line?

**Comment by dnfield** [comment]
128 is probably fine. We can always adjust if it's too aggressive or not aggressive enough.

**Comment by jacob314** [comment]
> 128 is probably fine. We can always adjust if it's too aggressive or not aggressive enough.
128K is probably fine for now.
It is more complex but maybe the threshold we want is the total wasted KB for the current screen. If you are rendering 10 images on the screen each one of which is wasting 50KB that is about as bad a single image wasting 500KB.

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 2]
```diff
@@ -338,6 +339,18 @@ class LoggingController {
         e.timestamp,
         summary: frameInfoText,
       ));
+    } else if (e.extensionKind == ImageSizesForFrame.eventName) {
+      final List<ImageSizesForFrame> images =
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart:343]: remove types that don't add value.

File: packages/devtools_app/lib/src/service_extensions.dart [hunk 3]
```diff
@@ -93,6 +93,18 @@ final debugAllowBanner = ToggleableServiceExtensionDescription<bool>._(
   gaItem: ga.debugBanner,
 );
 
+final invertOversizedImages = ToggleableServiceExtensionDescription<bool>._(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/service_extensions.dart:96]: It isn't great but material design has an invert colors icon.
https://material.io/resources/icons/?icon=invert_colors&style=baseline

File: packages/devtools_app/lib/src/console.dart (hunk 1 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/console.dart:81]: This was in order to get line wrapping behavior in the logging page's detail area. The flutter error's sent for image overage had very long lines and some of the error message was lost.

I can play around w/ how we display json in that view instead - I think the long line had some newlines embedded in it that we weren't respecting (that were json escaped). I don't think the details area gets very long generally, so I wasn't worried about virtualizing the display.

Comment (jacob314) [reply packages/devtools_app/lib/src/console.dart:81]: Ok. Virtualizing a list view in Flutter is trivial so it should just be the same lines of code but I agree the performance won't matter in this case unless we are displaying a really large block of text.

**Comment by devoncarew** [comment]
I have some updates to this PR in-flight, but for the location of the toggle button:

> add a toggle for the ext.flutter. invertOversizedImages service extension; this is on the Inspector page right now - it seems more immediately something people would use while looking at their app that a memory related toggle.

It's on the inspector page currently, and I don't see a good spot on the memory page for it. I think a better long term home for it would be:
- for us to add an 'Images' tab to the memory page
- for one of the available toggles in that new tab to be the 'oversized images' toggle

**Comment by jacob314** [comment]
I think an images tab in the memory page would be a good place for this.  I'd like the images tab to have this toggle along with a full list of all images currently displayed in the app.

File: packages/devtools_app/lib/src/service_extensions.dart (hunk 3 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/service_extensions.dart:97]: I was able to find a reasonable 'image' icon in the set of material ones.

**Comment by devoncarew** [comment]
This is updated w/ the latest changes now - ptal.

**Review by jacob314 — APPROVED** [review]
lgtm

**Pull request merged**